File: ejerciciosPracticos/python/ejercicio1.py

from datetime import datetime
from collections import defaultdict


ventas = [
    {"cliente": "Ana", "monto": "120.50", "fecha": "2026-01-20", "pais": "CO"},
    {"cliente": "Luis", "monto": "85", "fecha": "2026/01/21", "pais": "CO"},
    {"cliente": "Ana", "monto": None, "fecha": "2026-01-22", "pais": "MX"},
    {"cliente": "Maria", "monto": "210.00", "fecha": "22-01-2026", "pais": "CO"},
    {"cliente": "Luis", "monto": "error", "fecha": "2026-01-23", "pais": "CO"},
    {"cliente": "Carlos", "monto": "99.99", "fecha": "2026-01-23", "pais": None},
]

print(ventas)

formatos_fecha = [
    "%Y-%m-%d",
    "%Y/%m/%d",
    "%d-%m-%Y",
]

ventasLimpias = []

for venta in ventas:
    try:
        monto = float(venta['monto'])
    except(TypeError, ValueError):
        continue

    fechaNormalizada = None
    for fmt in formatos_fecha:
        try:
            fechaNormalizada = datetime.strptime(venta['fecha'], fmt).strftime('%Y-%m-%d')
            break
        except ValueError:
            pass

    if fechaNormalizada is None:
        continue

    pais = venta['pais'] if venta['pais'] is not None else 'DESCONOCIDO'

    ventasLimpias.append({
        'cliente': venta['cliente'],
        'monto': monto,
        'fecha': fechaNormalizada,
        'pais': pais
    })

print(ventasLimpias)


# Total de ventas válidas
total_ventas = sum(v["monto"] for v in ventasLimpias)

# Venta promedio
venta_promedio = total_ventas / len(ventasLimpias)

# Total por país
ventas_por_pais = defaultdict(float)
for v in ventasLimpias:
    ventas_por_pais[v["pais"]] += v["monto"]

# Cliente que más gastó
gasto_por_cliente = defaultdict(float)
for v in ventasLimpias:
    gasto_por_cliente[v["cliente"]] += v["monto"]

cliente_top = max(gasto_por_cliente, key=gasto_por_cliente.get)

File: ejerciciosPracticos/php/ejercicio1.php
<?php

    $ventas = [
        ["cliente" => "Ana", "monto" => "150.00", "pais" => "CO"],
        ["cliente" => "Luis", "monto" => "90", "pais" => "CO"],
        ["cliente" => "Ana", "monto" => null, "pais" => "MX"],
        ["cliente" => "Maria", "monto" => "200.50", "pais" => ""],
        ["cliente" => "Luis", "monto" => "error", "pais" => "CO"],
        ["cliente" => "Carlos", "monto" => "120", "pais" => null],
    ];

    foreach($ventas as $dato => $valor){
        echo $valor['cliente']. ' ' .$valor['monto']. ' ' .$valor['pais'].'<br>';
    };

    $ventasLimpias = [];

    foreach($ventas as $venta){
        //1. validar monto
        if(!is_numeric($venta['monto'])){
            continue; //descartamos la venta
        };
        
        
        $montoLimpio = floatval($venta['monto']);
        

        //Limpiar pais
        $paisLimpio = $venta['pais'];
        if($paisLimpio === '' || $paisLimpio === null){
            $paisLimpio = 'DESCONOCIDO';
        }

        //ARMAR VENTA LIMPIA
        $ventasLimpias[] = [
            'cliente' => $venta['cliente'],
            'monto' => $montoLimpio,
            'pais' => $paisLimpio
        ];
    };

    foreach($ventasLimpias as $dato => $valor){
        echo $valor['cliente']. ' ' .$valor['monto']. ' ' .$valor['pais'].'<br>';
    };

    //total ingresos
    $totalIngresos = 0;

    foreach($ventasLimpias as $venta){
        $totalIngresos += $venta['monto'];
    }

    echo 'El ingreso total fue de $'.$totalIngresos.'<br>';

    //promedio Ingresos
    $cantidadVentas = count($ventasLimpias);
    $PromedioIngresos = $cantidadVentas > 0 ? $totalIngresos/$cantidadVentas :0;

    echo 'El promedio total de los ingresos fue de $'.$PromedioIngresos.'<br>';

    //Total ingresos por pais
    $ingresosPorPais = [];

    foreach($ventasLimpias as $ventas){
        $pais = $ventas['pais'];

        if(!isset($ingresosPorPais[$pais])){
            $ingresosPorPais[$pais] = 0;
        }

        $ingresosPorPais[$pais] += $ventas['monto'];
    }

    foreach($ingresosPorPais as $clave => $valor){
        echo 'El ingreso del pais: '.$clave.' Es de $'.$valor.'<br>';
    }

    //cliente que mas dinero gasto
    $gastoPorCliente = [];

    foreach($ventasLimpias as $venta){
        $cliente = $venta['cliente'];

        if(!isset($gastoPorCliente[$cliente])){
            $gastoPorCliente[$cliente] = 0;
        }

        $gastoPorCliente[$cliente] += $venta['monto'];
    }

    //buscar el mayor
    $clienteTop = null;
    $montoMax = 0;

    foreach($gastoPorCliente as $cliente => $monto){
        if($monto > $montoMax){
            $montoMax = $monto;
            $clienteTop = $cliente;
        }
    }

    echo 'El cliente con mayor ingreso fue '.$clienteTop.' con un ingreso de $'.$montoMax;
    

    